Mobile CPU allocator.

AshkanAliabadi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36202 Enable torch_speed_benchmark to accept different memory formats.
* #35791 Enable stateless XNNPACK linear.
* #35790 Enable stateless XNNPACK convolutions.
* **#36032 Mobile CPU allocator.**

QNNPACK AND XNNPACK may out-of-bound access the input and / or output tensors. This is by-design, and chosen to make the implementation of micro-kernels both simpler and faster as a result of not having to individually handle the corner cases where the number of processed elements is not a multiple of SIMD register width.  This behavior will trigger ASAN though, and may result in a segfault if the accessed memory location just so happens to fall on a page the current process has no read access to.  Here we define a custom allocator that allocates the extra storage required to keep this behavior safe.  This allocator could have been restricted to QNNPACK and XNNPACK only, but that would have negative performance ramifications, as input tensors must now be reallocated, and copied over, if the tensor is not allocated with this allocator to begin with.  Making this allocator the default on mobile builds minimizes the probability of unnecessary reallocations and copies, and also enables acceleration of operations where the output tensor is allocated outside of the function doing the implementation, wherein the implementation cannot simply re-allocate the output with the guarding allocator.

Differential Revision: [D20970217](https://our.internmc.facebook.com/intern/diff/D20970217)

Commit:
Mobile CPU allocator.

[ghstack-poisoned]

Commit:
Update on "Mobile CPU allocator."




[ghstack-poisoned]

Commit:
Update on "Mobile CPU allocator."




[ghstack-poisoned]

kimishpatel commented on c10/core/CPUAllocator.cpp:
```diff
@@ -153,6 +153,8 @@ void SetCPUAllocator(at::Allocator* alloc) {
 }
 
+#ifndef C10_MOBILE
+
```
I think it will be better to have the ifdef inside of this function, instead of having a separate mobile specific def of this function. This function being `GetDefaultCPUAllocator`, no?

dreiss replied:
Makese sense.  I'm also a bit concerned about the `REGISTER_ALLOCATOR(DeviceType::CPU, &g_mobile_cpu_allocator);` in the other file because registering the allocator for a given device in different files is difficult to keep track of.  The change you're suggesting would eliminate that issue as well, right?

AshkanAliabadi replied:
g_mobile_cpu_allocator is not visible in this file.  The class that this is an instance of is not available here either.  I can put all implementation details into this file if that's what you guys prefer.

AshkanAliabadi replied:
(by which I mean removing MobileCPUAllocator and copying its contents here.)

dreiss replied:
I think moving the implementation is fine since it is small and has no other dependencies.  But you could also call GetDefaultMobileCPUAllocator from this file instead, right?

kimishpatel replied:
> g_mobile_cpu_allocator is not visible in this file. The class that this is an instance of is not available here either. I can put all implementation details into this file if that's what you guys prefer.

Is it possible to do that by including the definition in MobileAllocator.h and including that here? And also doing this `DefaultMobileCPUAllocator<std::max(8u, Configuration::kAlignment), 16u> g_mobile_cpu_allocator;` in this file inside `#ifdef C10_MOBILE`. That way static instance is created only for mobile build. Then we should be able to move ifdef inside `GetDefaultCPUAllocator` function. What do you think?

AshkanAliabadi replied:
g_mobile_cpu_allocator is not really mobile specific since QNNPACK and XNNPACK are not.  In other words, QNNPACK and XNNPACK can be used on non-mobile builds making the presence of this allocator a requirement.

AshkanAliabadi replied:
> I think moving the implementation is fine since it is small and has no other dependencies. But you could also call GetDefaultMobileCPUAllocator from this file instead, right?

Right you could call the function but not register the allocator since it's got internal linkage. Won't be an issue if I move the code. Let me know which you guys prefer.  No strong preference here.

kimishpatel replied:
> g_mobile_cpu_allocator is not really mobile specific since QNNPACK and XNNPACK are not. In other words, QNNPACK and XNNPACK can be used on non-mobile builds making the presence of this allocator a requirement.

I see.So can we still do `DefaultMobileCPUAllocator<std::max(8u, Configuration::kAlignment), 16u> g_mobile_cpu_allocator; ` outside of ifdef? Upto you. Just sharing what came to my mind.

dreiss replied:
I think moving the code is fine.  I think putting `g_mobile_cpu_allocator` outside of ifdef is also fine.  The only thing I don't think is fine is having `REGISTER_ALLOCATOR(DeviceType::CPU,` show up in two different files.

AshkanAliabadi replied:
Sounds good.  Will move everything into one file.

AshkanAliabadi replied:
Done. Marking as resolved.

Commit:
Update on "Mobile CPU allocator."




[ghstack-poisoned]

Commit:
Update on "Mobile CPU allocator."




[ghstack-poisoned]

Commit:
Update on "Mobile CPU allocator."




[ghstack-poisoned]

Commit:
Update on "Mobile CPU allocator."




[ghstack-poisoned]

Commit:
Update on "Mobile CPU allocator."




[ghstack-poisoned]

AshkanAliabadi commented:
I need to fix a few things but wanted to get feedback early.  I'll update the commit message accordingly.  The goal is two fold: 1) avoid unnecessary input reallocations and copies in XNNPACK. 2) enable usage of QNNPACK/XNNPACK for operators that take the output tensor as an argument, namely the operators ending in underscore, and ending in _out.  As a bonus, I am planning to expand this later on to do lazy deallocations to help with the comparatively expensive tensor deallocations we're seeing the profiles.

AshkanAliabadi commented on c10/core/CPUAllocator.cpp:
```diff
@@ -36,7 +36,7 @@ void memset_junk(void* data, size_t num) {
 }
 
-void* alloc_cpu(size_t nbytes) {
+void* alloc_cpu(size_t nbytes, const size_t alignment) {
```
Must either modify this function to accept alignment as input, or copy-paste it with this addition.  I decided to go with modifying this function, defaulting the second argument to gAlignment.  Any issues with this?  API stability and backwards compatibility, etc.?

dreiss replied:
It looks like this argument will always be 16 on mobile and 64 on non-mobile.  Maybe we should just change the definition of gAlignment to reflect that and then we don't need this part of the change?

AshkanAliabadi replied:
That'll push the #ifdef ARM to the header, are we OK with that?  Alternatively we can move gAlignment to the .cpp file.

AshkanAliabadi replied:
Done. Marking as resolved.

AshkanAliabadi commented on c10/core/CPUAllocator.cpp:
```diff
@@ -153,14 +153,18 @@ void SetCPUAllocator(at::Allocator* alloc) {
 
-// Global default CPU Allocator
-static DefaultCPUAllocator g_cpu_alloc;
+// #ifndef C10_MOBILE
```
Disregard this section for now.  I am intentionally making this path run for all mobile and non-mobile builds to exercise it more with CI.  This will be mobile only.

AshkanAliabadi commented on c10/core/MobileAllocator.cpp:
```diff
@@ -0,0 +1,93 @@
+  return &g_mobile_cpu_allocator;
+}
+
+// #ifdef C10_MOBILE
```
Will be mobile only.  Just exercising through CI to catch errors.

AshkanAliabadi commented on c10/core/MobileAllocator.cpp:
```diff
@@ -0,0 +1,93 @@
+  };
+};
+
+DefaultMobileCPUAllocator<std::max(8u, Configuration::kAlignment), 16u> g_mobile_cpu_allocator;
```
QNNPACK requires 8 bytes guard before while XNNPACK requires 16 bytes after.  The thing to keep in mind here is that memory needs to be aligned not at the point of allocation, but 8 bytes into the allocation because that's what the kernels see.  Basically we need the equivalent of (Microsoft's) [_aligned_offset_malloc](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/aligned-offset-malloc?view=vs-2019) on all platforms, which is not present so must emulate.

This means we have to paths moving forward: 1) either make the pre-padding equal to the alignment so when we step that much forward into the buffer we land on an aligned location, or 2) use malloc to achieve the same behavior as a result of allocating extra "offset + pre_guard" bytes and doing pointer arithmatic to return an aligned address.  I chose the former but we can do the latter.

dreiss replied:
This seems like a reasonable choice.  Can you leave a comment documenting where the constants 8 and 16 come from, and why we're doing the max?  You don't need to include the full context about MS's version and the alternative implementation.

AshkanAliabadi replied:
Done. Marking as resolved.

Commit:
Update on "Mobile CPU allocator."




[ghstack-poisoned]

Commit:
Update on "Mobile CPU allocator."




[ghstack-poisoned]

Commit:
Update on "Mobile CPU allocator."




[ghstack-poisoned]

dreiss commented on c10/core/MobileAllocator.cpp:
```diff
@@ -0,0 +1,106 @@
+  }
+
+  virtual DataPtr allocate(const size_t nbytes) const override {
+    if (C10_UNLIKELY(0u == nbytes)) {
+      return {
+        nullptr,
+        nullptr,
+        &deleter,
+        at::Device(DeviceType::CPU),
+      };
+    }
+
```
Any particular reason to special case this?  It increases code complexity and binary size.

AshkanAliabadi replied:
0 bytes would still allocate pre guard + post guard bytes when passed to alloc_cpu.  Fails a few tests that expect a nullptr returned on 0 byte allocations.

dreiss replied:
IMO those tests are broken because malloc is allowed to return non-NULL for zero-byte allocations, but okay.

dreiss commented on c10/core/MobileAllocator.cpp:
```diff
@@ -0,0 +1,106 @@
+  }
+
+ private:
+  union Cast final {
+    void * const as_void_ptr;
+    uint8_t * as_byte_ptr;
+  };
```
Why do we need this?  A few casts for the `+=` would reduce the number of concepts in this file and allow someone to fully understand that line without looking elsewhere for this definition.

AshkanAliabadi replied:
I personally find this easier to read and less error prone than casting.  Will switch to casts.  No strong preference.

AshkanAliabadi replied:
Done.

dreiss commented on c10/core/MobileAllocator.cpp:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    const Cast memory{pointer};
+    c10::free_cpu(memory.as_byte_ptr - PreGuardBytes);
```
It looks like `pointer` here is the context pointer, rather than the data pointer.  We could set that to the original allocated memory (not the offset value) and avoid the arithmetic here.

AshkanAliabadi replied:
Sorry I'm not following ... :/

dreiss replied:
In allocate, you construct a DataPtr with two fields, the pointer to the data, and the "context" pointer, which is what gets passed to free here.  You set both to the "allocated pointer plus pre-guard bytes" value.  But if you set the second one to the original allocated value, you could just pass that to free_cpu here.

AshkanAliabadi replied:
I see, nice!

AshkanAliabadi replied:
Done.

dreiss reviewed (changes requested):
I think this is a good strategy.

AshkanAliabadi commented:
Right you could call the function but not register the allocator since it's
got private linkage. Won't be an issue if I move the code. let me know
which you prefer.

On Wed, Apr 8, 2020, 1:54 PM David Reiss <notifications@github.com> wrote:

> *@dreiss* commented on this pull request.
> ------------------------------
>
> In c10/core/CPUAllocator.cpp
> <https://github.com/pytorch/pytorch/pull/36032#discussion_r405808525>:
>
> > @@ -153,6 +153,8 @@ void SetCPUAllocator(at::Allocator* alloc) {
>    SetAllocator(DeviceType::CPU, alloc);
>  }
>
> +#ifndef C10_MOBILE
> +
>
> I think moving the implementation is fine since it is small and has no
> other dependencies. But you could also call GetDefaultMobileCPUAllocator
> from this file instead, right?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/36032#discussion_r405808525>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD24NKBRPGQFUZ2JPKGOZPDRLTP7XANCNFSM4L7JELGQ>
> .
>

Commit:
Update on "Mobile CPU allocator."




[ghstack-poisoned]

Commit:
Update on "Mobile CPU allocator."


QNNPACK AND XNNPACK may out-of-bound access the input and / or output tensors. This is by-design, and chosen to make the implementation of micro-kernels both simpler and faster as a result of not having to individually handle the corner cases where the number of processed elements is not a multiple of SIMD register width.  This behavior will trigger ASAN though, and may result in a segfault if the accessed memory location just so happens to fall on a page the current process has no read access to.  Here we define a custom allocator that allocates the extra storage required to keep this behavior safe.  This allocator could have been restricted to QNNPACK and XNNPACK only, but that would have negative performance ramifications, as input tensors must now be reallocated, and copied over, if the tensor is not allocated with this allocator to begin with.  Making this allocator the default on mobile builds minimizes the probability of unnecessary reallocations and copies, and also enables acceleration of operations where the output tensor is allocated outside of the function doing the implementation, wherein the implementation cannot simply re-allocate the output with the guarding allocator.

Differential Revision: [D20970217](https://our.internmc.facebook.com/intern/diff/D20970217)

[ghstack-poisoned]

Commit:
Update on "Mobile CPU allocator."


QNNPACK AND XNNPACK may out-of-bound access the input and / or output tensors. This is by-design, and chosen to make the implementation of micro-kernels both simpler and faster as a result of not having to individually handle the corner cases where the number of processed elements is not a multiple of SIMD register width.  This behavior will trigger ASAN though, and may result in a segfault if the accessed memory location just so happens to fall on a page the current process has no read access to.  Here we define a custom allocator that allocates the extra storage required to keep this behavior safe.  This allocator could have been restricted to QNNPACK and XNNPACK only, but that would have negative performance ramifications, as input tensors must now be reallocated, and copied over, if the tensor is not allocated with this allocator to begin with.  Making this allocator the default on mobile builds minimizes the probability of unnecessary reallocations and copies, and also enables acceleration of operations where the output tensor is allocated outside of the function doing the implementation, wherein the implementation cannot simply re-allocate the output with the guarding allocator.

Differential Revision: [D20970217](https://our.internmc.facebook.com/intern/diff/D20970217)

[ghstack-poisoned]

dreiss commented on c10/core/CPUAllocator.cpp:
```diff
@@ -162,6 +246,8 @@ at::Allocator* GetDefaultCPUAllocator() {
 REGISTER_ALLOCATOR(DeviceType::CPU, &g_cpu_alloc);
 
+#endif /* C10_Mobile */
+
```
This section (the whole ifdef-else-endif block, I can't highlight the whole thing in github) is very easy to understand and should be easy to modify safely.  The part above the ifdef could possibly be moved to another .cpp file since `GetDefaultMobileCPUAllocator` is already declared in a header, but given how short it is, I don't think it's necessary.

Commit:
Update on "Mobile CPU allocator."


QNNPACK AND XNNPACK may out-of-bound access the input and / or output tensors. This is by-design, and chosen to make the implementation of micro-kernels both simpler and faster as a result of not having to individually handle the corner cases where the number of processed elements is not a multiple of SIMD register width.  This behavior will trigger ASAN though, and may result in a segfault if the accessed memory location just so happens to fall on a page the current process has no read access to.  Here we define a custom allocator that allocates the extra storage required to keep this behavior safe.  This allocator could have been restricted to QNNPACK and XNNPACK only, but that would have negative performance ramifications, as input tensors must now be reallocated, and copied over, if the tensor is not allocated with this allocator to begin with.  Making this allocator the default on mobile builds minimizes the probability of unnecessary reallocations and copies, and also enables acceleration of operations where the output tensor is allocated outside of the function doing the implementation, wherein the implementation cannot simply re-allocate the output with the guarding allocator.

Differential Revision: [D20970217](https://our.internmc.facebook.com/intern/diff/D20970217)

[ghstack-poisoned]

Commit:
Update on "Mobile CPU allocator."


QNNPACK AND XNNPACK may out-of-bound access the input and / or output tensors. This is by-design, and chosen to make the implementation of micro-kernels both simpler and faster as a result of not having to individually handle the corner cases where the number of processed elements is not a multiple of SIMD register width.  This behavior will trigger ASAN though, and may result in a segfault if the accessed memory location just so happens to fall on a page the current process has no read access to.  Here we define a custom allocator that allocates the extra storage required to keep this behavior safe.  This allocator could have been restricted to QNNPACK and XNNPACK only, but that would have negative performance ramifications, as input tensors must now be reallocated, and copied over, if the tensor is not allocated with this allocator to begin with.  Making this allocator the default on mobile builds minimizes the probability of unnecessary reallocations and copies, and also enables acceleration of operations where the output tensor is allocated outside of the function doing the implementation, wherein the implementation cannot simply re-allocate the output with the guarding allocator.

Differential Revision: [D20970217](https://our.internmc.facebook.com/intern/diff/D20970217)

[ghstack-poisoned]

Commit:
Update on "Mobile CPU allocator."


QNNPACK AND XNNPACK may out-of-bound access the input and / or output tensors. This is by-design, and chosen to make the implementation of micro-kernels both simpler and faster as a result of not having to individually handle the corner cases where the number of processed elements is not a multiple of SIMD register width.  This behavior will trigger ASAN though, and may result in a segfault if the accessed memory location just so happens to fall on a page the current process has no read access to.  Here we define a custom allocator that allocates the extra storage required to keep this behavior safe.  This allocator could have been restricted to QNNPACK and XNNPACK only, but that would have negative performance ramifications, as input tensors must now be reallocated, and copied over, if the tensor is not allocated with this allocator to begin with.  Making this allocator the default on mobile builds minimizes the probability of unnecessary reallocations and copies, and also enables acceleration of operations where the output tensor is allocated outside of the function doing the implementation, wherein the implementation cannot simply re-allocate the output with the guarding allocator.

Differential Revision: [D20970217](https://our.internmc.facebook.com/intern/diff/D20970217)

[ghstack-poisoned]

Commit:
Update on "Mobile CPU allocator."


QNNPACK AND XNNPACK may out-of-bound access the input and / or output tensors. This is by-design, and chosen to make the implementation of micro-kernels both simpler and faster as a result of not having to individually handle the corner cases where the number of processed elements is not a multiple of SIMD register width.  This behavior will trigger ASAN though, and may result in a segfault if the accessed memory location just so happens to fall on a page the current process has no read access to.  Here we define a custom allocator that allocates the extra storage required to keep this behavior safe.  This allocator could have been restricted to QNNPACK and XNNPACK only, but that would have negative performance ramifications, as input tensors must now be reallocated, and copied over, if the tensor is not allocated with this allocator to begin with.  Making this allocator the default on mobile builds minimizes the probability of unnecessary reallocations and copies, and also enables acceleration of operations where the output tensor is allocated outside of the function doing the implementation, wherein the implementation cannot simply re-allocate the output with the guarding allocator.

Differential Revision: [D20970217](https://our.internmc.facebook.com/intern/diff/D20970217)

[ghstack-poisoned]

Commit:
Update on "Mobile CPU allocator."


QNNPACK AND XNNPACK may out-of-bound access the input and / or output tensors. This is by-design, and chosen to make the implementation of micro-kernels both simpler and faster as a result of not having to individually handle the corner cases where the number of processed elements is not a multiple of SIMD register width.  This behavior will trigger ASAN though, and may result in a segfault if the accessed memory location just so happens to fall on a page the current process has no read access to.  Here we define a custom allocator that allocates the extra storage required to keep this behavior safe.  This allocator could have been restricted to QNNPACK and XNNPACK only, but that would have negative performance ramifications, as input tensors must now be reallocated, and copied over, if the tensor is not allocated with this allocator to begin with.  Making this allocator the default on mobile builds minimizes the probability of unnecessary reallocations and copies, and also enables acceleration of operations where the output tensor is allocated outside of the function doing the implementation, wherein the implementation cannot simply re-allocate the output with the guarding allocator.

Differential Revision: [D20970217](https://our.internmc.facebook.com/intern/diff/D20970217)

[ghstack-poisoned]

Commit:
Update on "Mobile CPU allocator."


QNNPACK AND XNNPACK may out-of-bound access the input and / or output tensors. This is by-design, and chosen to make the implementation of micro-kernels both simpler and faster as a result of not having to individually handle the corner cases where the number of processed elements is not a multiple of SIMD register width.  This behavior will trigger ASAN though, and may result in a segfault if the accessed memory location just so happens to fall on a page the current process has no read access to.  Here we define a custom allocator that allocates the extra storage required to keep this behavior safe.  This allocator could have been restricted to QNNPACK and XNNPACK only, but that would have negative performance ramifications, as input tensors must now be reallocated, and copied over, if the tensor is not allocated with this allocator to begin with.  Making this allocator the default on mobile builds minimizes the probability of unnecessary reallocations and copies, and also enables acceleration of operations where the output tensor is allocated outside of the function doing the implementation, wherein the implementation cannot simply re-allocate the output with the guarding allocator.

Differential Revision: [D20970217](https://our.internmc.facebook.com/intern/diff/D20970217)

[ghstack-poisoned]
